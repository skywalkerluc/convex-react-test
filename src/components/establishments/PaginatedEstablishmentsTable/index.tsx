import { useState } from 'react';
import { useEstablishments } from '../../../hooks/useEstablishments';
import EstablishmentsTable from '../EstablishmentsTable';
import EstablishmentsTableNavigation from '../EstablishmentsTableNavigation';
import { PaginationParams } from '../../../types/establishment';
import styles from './styles.module.css';

const PaginatedEstablishmentsTable = ()  => {
  const [pagination, setPagination] = useState<PaginationParams>({ page: 1, pageSize: 10 });
  const { establishments, totalPages, loading, error } = useEstablishments(pagination);

  const handlePrevious = () => {
    setPagination(prev => ({
      ...prev,
      page: Math.max(1, prev.page - 1)
    }));
  };

  const handleNext = () => {
    setPagination(prev => ({
      ...prev,
      page: Math.min(totalPages, prev.page + 1)
    }));
  };

  return (
    <div className={styles.container}>
      {error && (
        <div className={styles.error} role="alert">
          {error}
        </div>
      )}

      <div className={styles.tableContainer}>
        <EstablishmentsTable establishments={establishments} />
      </div>

      <EstablishmentsTableNavigation
        currentPage={pagination.page}
        totalPages={totalPages}
        onPrevious={handlePrevious}
        onNext={handleNext}
        loading={loading}
      />
    </div>
  );
};

export default PaginatedEstablishmentsTable;