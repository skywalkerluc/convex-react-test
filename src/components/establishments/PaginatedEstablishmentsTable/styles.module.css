.container {
    max-width: 1200px;
    margin: 2rem auto;
    padding: 0 1rem;
  }
  
  .tableContainer {
    min-height: 500px;
    position: relative;
  }
  
  .error {
    color: #dc3545;
    background-color: #f8d7da;
    padding: 1rem;
    margin-bottom: 1rem;
    border-radius: 4px;
    border: 1px solid #f5c6cb;
  }