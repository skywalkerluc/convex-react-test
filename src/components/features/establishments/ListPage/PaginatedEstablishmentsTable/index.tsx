import { useState } from 'react';

import { useEstablishments } from '../../../../../hooks/useEstablishments';
import { PaginationParams } from '../../../../../types/pagination';
import { AuthorityFilter } from '../../../filters/AuthorityFilter';
import EstablishmentsTable from '../EstablishmentsTable';
import EstablishmentsTableNavigation from '../EstablishmentsTableNavigation';

import styles from './styles.module.css';

const PaginatedEstablishmentsTable = () => {
  const [options, setOptions] = useState<PaginationParams>({ page: 1, pageSize: 5 });
  const { establishments, totalPages, loading, error } = useEstablishments(options);

  const handlePrevious = () => {
    setOptions((prev) => ({
      ...prev,
      page: Math.max(1, prev.page - 1),
    }));
  };

  const handleNext = () => {
    setOptions((prev) => ({
      ...prev,
      page: Math.min(totalPages, prev.page + 1),
    }));
  };

  const handleAuthorityChange = (authority: string) => {
    setOptions((prev) => ({ ...prev, authority, page: 1 }));
  };

  const isAbortError = (error: Error | null): boolean => {
    return error instanceof DOMException && error.name === 'AbortError';
  };

  return (
    <div className={styles.container}>
      <AuthorityFilter value={options.authority || ''} onChange={handleAuthorityChange} />
      {error && !isAbortError(error) && (
        <div className={styles.error} role='alert'>
          {error.message}
        </div>
      )}

      <div className={styles.tableWrapper}>
        <EstablishmentsTable establishments={establishments} loading={loading} />
      </div>

      <EstablishmentsTableNavigation
        currentPage={options.page}
        totalPages={totalPages}
        onPrevious={handlePrevious}
        onNext={handleNext}
        loading={loading}
      />
    </div>
  );
};

export default PaginatedEstablishmentsTable;
