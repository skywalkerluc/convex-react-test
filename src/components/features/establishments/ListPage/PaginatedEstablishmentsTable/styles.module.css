.container {
    max-width: 1200px;
    margin: 2rem auto;
    padding: 0 1rem;
  }
  
  .tableWrapper {
    min-height: 300px;
    position: relative;
  }

  .tableWrapper > table {
    width: 100%;
    border-collapse: collapse;
  }
  
  .tableWrapper > table thead {
    background: #f8f9fa;
    box-shadow: 0 2px 4px rgba(0,0,0,0.05);
  }