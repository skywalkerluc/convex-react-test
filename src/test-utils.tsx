import { ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';

export const createWrapper = (options?: { router?: { initialEntries?: string[] } }) => {
  const Wrapper = ({ children }: { children: ReactElement }) => (
    <MemoryRouter initialEntries={options?.router?.initialEntries || ['/']}>
      {children}
    </MemoryRouter>
  );
  return Wrapper;
};
