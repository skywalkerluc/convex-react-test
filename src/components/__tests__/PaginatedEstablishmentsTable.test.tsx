import { render, screen } from '@testing-library/react';

import { createWrapper } from '../../test-utils';
import PaginatedEstablishmentsTable from '../features/establishments/ListPage/PaginatedEstablishmentsTable';

jest.mock('../../context/AppContext', () => ({
  useAppContext: () => ({
    state: {
      establishments: [
        {
          id: 1,
          businessName: 'Test Restaurant',
          ratingValue: '5',
          addressLine1: 'Test Street 1',
          postCode: 'TE5 5TP',
        },
      ],
      authorities: [],
      totalPages: 1,
      pagination: { page: 1, pageSize: 5 },
      isLoading: false,
      error: null,
    },
    dispatch: jest.fn(),
  }),
}));

describe('PaginatedEstablishmentsTable', () => {
  it('should render establishments', () => {
    render(<PaginatedEstablishmentsTable />, { wrapper: createWrapper() });
    expect(screen.getByText('Test Restaurant')).toBeInTheDocument();
    expect(screen.getByText('5')).toBeInTheDocument();
  });
});
