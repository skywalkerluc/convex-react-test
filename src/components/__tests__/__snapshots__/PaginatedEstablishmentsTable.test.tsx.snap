// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatedEstablishmentsTable matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="tableContainer"
    >
      <table
        aria-label="Tabela de estabelecimentos"
        class="table"
      >
        <thead>
          <tr>
            <th
              class="header"
            >
              Business Name
            </th>
            <th
              class="header"
            >
              Rating
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
    <div
      class="navigationContainer"
      data-testid="pagination-controls"
    >
      <button
        aria-disabled="true"
        aria-label="Previous page"
        class="button"
        disabled=""
        type="button"
      >
        Previous
      </button>
      <span
        class="pageInfo"
      >
        Page 1 of 5
      </span>
      <button
        aria-disabled="false"
        aria-label="Next page"
        class="button"
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PaginatedEstablishmentsTable matches snapshot with error 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="error"
      role="alert"
    >
      Erro de conex√£o
    </div>
    <div
      class="tableContainer"
    >
      <table
        aria-label="Tabela de estabelecimentos"
        class="table"
      >
        <thead>
          <tr>
            <th
              class="header"
            >
              Business Name
            </th>
            <th
              class="header"
            >
              Rating
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
    <div
      class="navigationContainer"
      data-testid="pagination-controls"
    >
      <button
        aria-disabled="true"
        aria-label="Previous page"
        class="button"
        disabled=""
        type="button"
      >
        Previous
      </button>
      <span
        class="pageInfo"
      >
        Page 1 of 0
      </span>
      <button
        aria-disabled="false"
        aria-label="Next page"
        class="button"
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</DocumentFragment>
`;
